pkg/cli/cert/cert.go

```diff
@@ -0,0 +1,237 @@
+package cert
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/erikdubbelboer/gspt"
+	"github.com/otiai10/copy"
+	"github.com/rancher/k3s/pkg/cli/cmds"
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/daemons/control/deps"
+	"github.com/rancher/k3s/pkg/datadir"
+	"github.com/rancher/k3s/pkg/server"
+	"github.com/rancher/k3s/pkg/version"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+const (
+	adminComponent             = "admin"
+	apiServerComponent         = "api-server"
+	controllerManagerComponent = "controller-manager"
+	schedulerComponent         = "scheduler"
+	etcdComponent              = "etcd"
+	programControllerComponent = "-controller"
+	authProxyComponent         = "auth-proxy"
+	cloudControllerComponent   = "cloud-controller"
+	kubeletComponent           = "kubelet"
+	kubeProxyComponent         = "kube-proxy"
+)
+
+func commandSetup(app *cli.Context, cfg *cmds.Server, sc *server.Config) (string, string, error) {
+	gspt.SetProcTitle(os.Args[0])
+
+	nodeName := app.String("node-name")
+	if nodeName == "" {
+		h, err := os.Hostname()
+		if err != nil {
+			return "", "", err
+		}
+		nodeName = h
+	}
+
+	os.Setenv("NODE_NAME", nodeName)
+
+	sc.ControlConfig.DataDir = cfg.DataDir
+	sc.ControlConfig.Runtime = &config.ControlRuntime{}
+	dataDir, err := datadir.Resolve(cfg.DataDir)
+	if err != nil {
+		return "", "", err
+	}
+	return filepath.Join(dataDir, "server"), filepath.Join(dataDir, "agent"), err
+}
+
+func Run(app *cli.Context) error {
+	if err := cmds.InitLogging(); err != nil {
+		return err
+	}
+	return rotate(app, &cmds.ServerConfig)
+}
+
+func rotate(app *cli.Context, cfg *cmds.Server) error {
+	var serverConfig server.Config
+
+	serverDataDir, agentDataDir, err := commandSetup(app, cfg, &serverConfig)
+	if err != nil {
+		return err
+	}
+
+	serverConfig.ControlConfig.DataDir = serverDataDir
+	serverConfig.ControlConfig.Runtime = &config.ControlRuntime{}
+	deps.CreateRuntimeCertFiles(&serverConfig.ControlConfig, serverConfig.ControlConfig.Runtime)
+
+	tlsBackupDir, err := backupCertificates(serverDataDir, agentDataDir)
+	if err != nil {
+		return err
+	}
+
+	if len(cmds.ComponentList) == 0 {
+		// rotate all certs
+		logrus.Infof("Rotating certificates for all services")
+		return rotateAllCerts(tlsBackupDir, filepath.Join(serverDataDir, "tls"), agentDataDir)
```

Comment:
Rather than having an empty list call the special `rotateAllCerts` function that has a lot of the same logic as the for loop below, could we (if the list is empty):
1. try to determine if the local node is a server or agent
2. populate the appropriate set of defaults into the ComponentList
3. let the for loop below handle the rotation for all the components
4. get rid of the rotateAllCerts function, because the for loop covers everything

Reply:
yup sounds good